### PR TITLE
[RW-9664][risk=no] [UI][pre-prod] Open RStudio & SAS in AoU iframe to avoid timeouts  <- For SAS

### PR DESCRIPTION
**Pull request by NehaBroad**
All functionality works similar to that of RStudio

https://github.com/all-of-us/workbench/assets/34481816/38c8ff90-c6d4-47e8-87e6-ee5c4d776ce3

---
**PR checklist**

- [x] I have included an issue ID or "no ticket" in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have included a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have manually tested this change and my testing process is described above.
- [ ] This change includes appropriate automated tests, and I have documented any behavior that cannot be tested with code.
- [x] I have added explanatory comments where the logic is not obvious.
- One or more of the following is true:
  - [x] This change is intended to complete a JIRA story, so I have checked that all AC are met for that story.
  - [ ] This change fixes a bug, so I have ensured the steps to reproduce are in the Jira ticket or provided above.
  - [ ] This change impacts deployment safety (e.g. removing/altering APIs which are in use), so I have documented the impacts in the description.
  - [ ] This change includes a new feature flag, so I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later.
  - [ ] This change modifies the UI, so I have taken screenshots or recordings of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P).
  - [ ] This change modifies API behavior, so I have run the relevant E2E tests locally because API changes are not covered by our PR checks.
  - [ ] None of the above apply to this change.

### CODE REVIEW

File: ui/src/app/routing/guards.spec.tsx [hunk 1]
```diff
@@ -222,4 +240,23 @@ describe('redirectTo', () => {
   ])('%s', (desc, expected: string, profile: Profile) => {
     expect(shouldRedirectToMaybe(profile)).toEqual(expected);
   });
+
+  it('Should verify App defined in URL is valid', async () => {
```
Comment (jmthibault79) [inline ui/src/app/routing/guards.spec.tsx:244]: good test, thanks!

However it's really two tests, so let's split it up

File: ui/src/app/routing/guards.spec.tsx [hunk 2]
```diff
@@ -222,4 +240,23 @@ describe('redirectTo', () => {
...
+    const appTypes = Object.values(UIAppType);
+
+    // All URLs containing "App" in UIAppType are considered valid and require no redirection.
+    let appIsValid = false;
+    appTypes.forEach((appType: UIAppType) => {
+      const pathToRstudioApp = `${hostPath}/workspaces/ws/id/${analysisTabName}/userApp/${appType}`;
```
Comment (jmthibault79) [inline ui/src/app/routing/guards.spec.tsx:250]: rename - not always RStudio

File: ui/src/app/pages/appAnalysis/app-selector.tsx [hunk 3]
```diff
@@ -72,7 +72,12 @@ export const AppSelector = (props: AppSelectorProps) => {
         break;
       case UIAppType.SAS:
         setVisibleModal(VisibleModal.None);
-        openSASOrConfigPanel(workspace.namespace, userApps);
+        openSASOrConfigPanel(
```
Comment (jmthibault79) [inline ui/src/app/pages/appAnalysis/app-selector.tsx:75]: we can collapse the RStudio and SAS cases if we make this call fully generic.  that means the check can simply be `if (jupyter) doJupyter else doUserApp`

File: ui/src/app/pages/analysis/interactive-notebook.tsx [hunk 4]
```diff
@@ -539,18 +539,18 @@ export const InteractiveNotebook = fp.flow(
...
           if (appType === UIAppType.RSTUDIO) {
-            const {
-              userAppsStore: { userApps },
-              navigate,
-            } = this.props;
             openRStudioOrConfigPanel(ns, wsid, userApps, navigate);
           } else if (appType === UIAppType.SAS) {
-            const { userApps } = this.props.userAppsStore;
-            openSASOrConfigPanel(ns, userApps);
+            openSASOrConfigPanel(ns, wsid, userApps, navigate);
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/interactive-notebook.tsx:553]: we can also collapse this logic to 2 branches - Jupyter or User Apps

File: ui/src/app/pages/analysis/interactive-notebook.spec.tsx [hunk 5]
```diff
@@ -117,22 +116,11 @@ test.each([
...
     const editButton = screen.getByTitle('Edit');
     await user.click(editButton);
-    if (appType === 'RStudio') {
-      await waitFor(() => {
-        expect(mockNavigate).toHaveBeenCalledWith([
-          appDisplayPath('sampleNameSpace', 'sampleWorkspace', appType),
-        ]);
-      });
-    } else if (appType === 'SAS') {
-      // This is temp. Once SAS moves to iframe this if else will be removed
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/interactive-notebook.spec.tsx:127]: hooray!

File: ui/src/app/pages/analysis/gke-app-launcher.spec.tsx [hunk 6]
```diff
@@ -84,18 +91,67 @@ const setupEmptyUserStore = () => {
...
-it('Should display iframe with rStudio App Url', async () => {
-  setup();
-  // Confirm IFrame exist
-  expect(screen.getByTitle('RStudio embed')).toBeInTheDocument();
-  // Confirm IFrame source is same as proxyUrl for Rstudio L24
-  expect(screen.getByTitle('RStudio embed').getAttribute('src')).toBe(
-    RSTUDIO_FAKE_URL
-  );
-});
+test.each([
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/gke-app-launcher.spec.tsx:94]: optional: can use a `describe.each()` block instead, with the two tests inside it

File: ui/src/app/pages/analysis/gke-app-launcher.spec.tsx [hunk 7]
```diff
@@ -84,18 +91,67 @@ const setupEmptyUserStore = () => {
...
+  [UIAppType.RSTUDIO, RSTUDIO_FAKE_URL],
+  [UIAppType.SAS, SAS_FAKE_URL],
+])(
+  'Should display iframe with %s App Url',
+  async (app: UIAppType, iframeSrcUrl: string) => {
+    setup(app);
+    // Confirm IFrame exist
+    expect(screen.getByTitle(`${app} embed`)).toBeInTheDocument();
+    // Confirm IFrame source is same as proxyUrl for Rstudio
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/gke-app-launcher.spec.tsx:103]: remove RStudio

File: ui/src/app/pages/analysis/gke-app-launcher.spec.tsx [hunk 8]
```diff
@@ -84,18 +91,67 @@ const setupEmptyUserStore = () => {
...
+  }
+);
+
+test.each([
+  [UIAppType.RSTUDIO, 'rstudioApp', RSTUDIO_FAKE_URL],
+  [UIAppType.SAS, 'sasApp', SAS_FAKE_URL],
+])(
+  'Should redirect user to analysis page in case %s App is deleted',
+  async (app: UIAppType, name: string, iframeUrl: string) => {
+    // Confirm the RStudio app is up and displayed in iframe and no navigate is called
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/gke-app-launcher.spec.tsx:116]: remove RStudio

File: ui/src/app/components/apps-panel/utils.tsx [hunk 9]
```diff
@@ -139,6 +143,14 @@ export const toUIAppType: Record<AppType, UIAppType> = {
   [AppType.SAS]: UIAppType.SAS,
 };
 
+export const helpSidebarConfigIdForUIApp = (appType: UIAppType) => {
```
[the remark this answers is not in the document]
Comment (NehaBroad) [reply ui/src/app/components/apps-panel/utils.tsx:146]: This is cool

File: ui/src/app/utils/user-apps-utils.tsx [hunk 10]
```diff
@@ -142,40 +143,63 @@ export const openSAS = (
...
   if (userApp?.status === AppStatus.RUNNING) {
-    openRStudio(workspaceNamespace, workspaceId, userApp, navigate);
+    openAppInIframe(workspaceNamespace, workspaceId, userApp, navigate);
   } else {
-    setSidebarActiveIconStore.next(rstudioConfigIconId);
+    setSidebarActiveIconStore.next(helpSidebarConfigIdForUIApp(requestedApp));
   }
 };
 
+export const openRStudioOrConfigPanel = (
```
[the remark this answers is not in the document]
Comment (NehaBroad) [reply ui/src/app/utils/user-apps-utils.tsx:177]: We could pass an argument for appType however for now i like the function name to ensure exactly which App is being open. Once you are back, i would like to discuss how we can even update other class as well and refactor code.

File: ui/src/app/pages/appAnalysis/app-selector.tsx (hunk 3 above)
Comment (NehaBroad) [reply ui/src/app/pages/appAnalysis/app-selector.tsx:75]: Yes, there are other places as well that we can make generic. I would like to discuss with you once you are back and have a PR to do that. 
I am thinking apart from this file we could also generalize  RStudioButtonRow/SasButtonRow etc

File: ui/src/app/pages/analysis/interactive-notebook.tsx (hunk 4 above)
Comment (NehaBroad) [reply ui/src/app/pages/analysis/interactive-notebook.tsx:553]: done

File: ui/src/app/utils/user-apps-utils.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (NehaBroad) [reply ui/src/app/utils/user-apps-utils.tsx:177]: Changed my mind, looking at the ifs or case , removed this and Sas specific method

File: ui/src/app/pages/analysis/gke-app-launcher.spec.tsx (hunk 6 above)
Comment (NehaBroad) [reply ui/src/app/pages/analysis/gke-app-launcher.spec.tsx:94]: there are other test as well that doesnt require test.each so i am going to leave it like this. We can discuss more on monday

Comment (jmthibault79) [reply ui/src/app/pages/analysis/gke-app-launcher.spec.tsx:94]: no need, fine as-is

File: ui/src/app/pages/analysis/interactive-notebook.spec.tsx [hunk 11]
```diff
@@ -49,6 +52,8 @@ beforeEach(async () => {
   jest.mock('app/services/swagger-fetch-clients');
   mockAppsApi = jest.spyOn(swaggerClients, 'appsApi');
   mockNotebooksApi = jest.spyOn(swaggerClients, 'notebooksApi');
+  helpSidebarConfigIdForUIApp[UIAppType.RSTUDIO] = rstudioConfigIconId;
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/interactive-notebook.spec.tsx:55]: hmm.  We shouldn't need to do this.  There may be something wrong about how we're using the record here.  I'll think about this a bit.

Comment (jmthibault79) [reply ui/src/app/pages/analysis/interactive-notebook.spec.tsx:55]: I made https://github.com/all-of-us/workbench/pull/8232 to help look into this ... certainly low priority

**Pull request merged**